compiler/rustc_macros/src/diagnostics/diagnostic_builder.rs

```diff
@@ -156,16 +156,20 @@ impl DiagnosticDeriveBuilder {
         let name = name.as_str();
         let meta = attr.parse_meta()?;
 
-        let is_help_or_note = matches!(name, "help" | "note");
+        let is_help_note_or_warn = matches!(name, "help" | "note" | "warn_");
```

Comment:
So why `warn_` and not `warning`?

Reply:
`warning` is already used as a type-level attribute, e.g. `#[warning(passes::some_fluent_attr)]` like `#[error(..)]`. It could be used but it makes everything more complicated, particularly because `#[warn_]` can also be at the type-level with no meta items for a spanless warning, so you could have `#[error(..)] #[warning] struct X;` which would normally be an error - this type can't represent both an error and a warning - but that we'd actually need to understand was an error with an attached warning. Doable, just more work.